File: apps/demo-be-golang-gin/internal/router/router.go
// Package router configures the Gin HTTP router with all API routes and middleware
// for the demo-be-golang-gin application.
package router

import (
	"github.com/gin-gonic/gin"

	"github.com/wahidyankf/open-sharia-enterprise/apps/demo-be-golang-gin/internal/auth"
	"github.com/wahidyankf/open-sharia-enterprise/apps/demo-be-golang-gin/internal/config"
	"github.com/wahidyankf/open-sharia-enterprise/apps/demo-be-golang-gin/internal/handler"
	"github.com/wahidyankf/open-sharia-enterprise/apps/demo-be-golang-gin/internal/store"
)

// NewRouter builds and returns the Gin engine with all routes registered.
func NewRouter(st store.Store, jwtSvc *auth.JWTService, cfg *config.Config) *gin.Engine {
	r := gin.New()
	r.Use(gin.Recovery())

	h := handler.New(st, jwtSvc)

	r.GET("/health", h.Health)
	r.GET("/.well-known/jwks.json", h.JWKS)

	api := r.Group("/api/v1")

	authGroup := api.Group("/auth")
	authGroup.POST("/register", h.Register)
	authGroup.POST("/login", h.Login)
	authGroup.POST("/logout", h.Logout)
	authGroup.POST("/refresh", h.Refresh)
	authGroup.POST("/logout-all", auth.JWTMiddleware(jwtSvc, st), h.LogoutAll)

	users := api.Group("/users", auth.JWTMiddleware(jwtSvc, st))
	users.GET("/me", h.GetProfile)
	users.PATCH("/me", h.UpdateProfile)
	users.POST("/me/password", h.ChangePassword)
	users.POST("/me/deactivate", h.Deactivate)

	admin := api.Group("/admin", auth.JWTMiddleware(jwtSvc, st), auth.AdminMiddleware())
	admin.GET("/users", h.ListUsers)
	admin.POST("/users/:id/disable", h.DisableUser)
	admin.POST("/users/:id/enable", h.EnableUser)
	admin.POST("/users/:id/unlock", h.UnlockUser)
	admin.POST("/users/:id/force-password-reset", h.ForcePasswordReset)

	expenses := api.Group("/expenses", auth.JWTMiddleware(jwtSvc, st))
	expenses.POST("", h.CreateExpense)
	expenses.GET("", h.ListExpenses)
	expenses.GET("/summary", h.ExpenseSummary)
	expenses.GET("/:id", h.GetExpense)
	expenses.PUT("/:id", h.UpdateExpense)
	expenses.DELETE("/:id", h.DeleteExpense)
	expenses.POST("/:id/attachments", h.UploadAttachment)
	expenses.GET("/:id/attachments", h.ListAttachments)
	expenses.DELETE("/:id/attachments/:aid", h.DeleteAttachment)

	tokens := api.Group("/tokens", auth.JWTMiddleware(jwtSvc, st))
	tokens.GET("/claims", h.TokenClaims)

	reports := api.Group("/reports", auth.JWTMiddleware(jwtSvc, st))
	reports.GET("/pl", h.PLReport)

	if cfg.EnableTestAPI {
		testGroup := api.Group("/test")
		testGroup.POST("/reset-db", h.ResetDB)
		testGroup.POST("/promote-admin", h.PromoteAdmin)
	}

	return r
}

File: apps/a-demo-be-elixir-phoenix/lib/a_demo_be_exph_web/router.ex
defmodule AADemoBeExphWeb.Router do
  use AADemoBeExphWeb, :router

  pipeline :api do
    plug :accepts, ["json"]
    plug AADemoBeExphWeb.CorsPlug
  end

  pipeline :auth do
    plug Guardian.Plug.Pipeline,
      module: ADemoBeExph.Auth.Guardian,
      error_handler: AADemoBeExphWeb.AuthErrorHandler

    plug Guardian.Plug.VerifyHeader, scheme: "Bearer"
    plug Guardian.Plug.EnsureAuthenticated
    plug AADemoBeExphWeb.Plugs.CheckRevoked
    plug AADemoBeExphWeb.Plugs.CheckUserActive
  end

  scope "/" do
    get "/health", AADemoBeExphWeb.HealthController, :index
    get "/.well-known/jwks.json", AADemoBeExphWeb.JwksController, :index
  end

  if System.get_env("ENABLE_TEST_API") == "true" do
    scope "/api/v1/test", AADemoBeExphWeb do
      pipe_through :api

      post "/reset-db", TestApiController, :reset_db
      post "/promote-admin", TestApiController, :promote_admin
    end
  end

  scope "/api/v1", AADemoBeExphWeb do
    pipe_through :api

    scope "/auth" do
      post "/register", AuthController, :register
      post "/login", AuthController, :login
      post "/refresh", AuthController, :refresh
      post "/logout", AuthController, :logout
      post "/logout-all", AuthController, :logout_all
    end

    scope "/" do
      pipe_through :auth

      scope "/users" do
        get "/me", UserController, :me
        patch "/me", UserController, :update_me
        post "/me/password", UserController, :change_password
        post "/me/deactivate", UserController, :deactivate
      end

      scope "/admin" do
        get "/users", AdminController, :list_users
        post "/users/:id/disable", AdminController, :disable_user
        post "/users/:id/enable", AdminController, :enable_user
        post "/users/:id/unlock", AdminController, :unlock_user
        post "/users/:id/force-password-reset", AdminController, :force_password_reset
      end

      scope "/expenses" do
        get "/", ExpenseController, :index
        post "/", ExpenseController, :create
        get "/summary", ExpenseController, :summary
        get "/:id", ExpenseController, :show
        put "/:id", ExpenseController, :update
        delete "/:id", ExpenseController, :delete

        scope "/:expense_id/attachments" do
          get "/", AttachmentController, :index
          post "/", AttachmentController, :create
          get "/:att_id", AttachmentController, :show
          delete "/:att_id", AttachmentController, :delete
        end
      end

      scope "/reports" do
        get "/pl", ReportController, :pl
      end
    end
  end
end
